numpy/core/src/multiarray/datetime.c

```diff
@@ -1622,6 +1622,13 @@ compute_datetime_metadata_greatest_common_divisor(
 
     return 0;
 
+/*
+ * Note that the following errors do not use "InvalidPromotion", because they
+ * should promote (two datetimes should have a common datetime), but cannot.
+ * The promotion is thus valid, but fails.
+ * This is important, because `arr == arr` for these cannot reasonably return
+ * all ``False`` values.
```

Comment:
Does it fail instead? Maybe extend sentence "... values, but should fail instead."?

Reply:
Ah, should expand on this... The cast normally fails because it is considered "unsafe".  The old hack is unnecessary because we now find the loop directly in either case.